### PR TITLE
Debug parameter now sets debugpreference to continue

### PR DESCRIPTION
**Pull request by KirkMunro**
## PR Summary

This is bound to generate some discussion because it is a breaking change (but not to scripts -- it breaks current interactive behavior). With this PR in place, users who invoke a command with -Debug will no longer be presented with a prompt asking them if they want to enter a nested prompt, continue execution, or halt execution entirely. Instead, any messages sent to the debug stream will simply be sent to the debug stream and the script will continue execution. That is why this is identified as a breaking change, but only for interactive use.

As described in #7122, the current -Debug behavior dates back to PowerShell v1 and hasn't been useful since the debugger was introduced in PowerShell 2. It's time this functionality be made useful again, for all of the reasons identified in #7122 (so that we can actually have debug output from specific commands show up when they are invoked with -Debug _without_ having to go through a bunch of unwanted prompts). See the details in the bug report to understand why this change is desired by members of the PowerShell Community.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/issues/3255
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/System.Management.Automation/engine/MshCommandRuntime.cs [hunk 1]
```diff
@@ -2924,23 +2924,7 @@ internal ActionPreference DebugPreference
...
-                        }
-                        else
-                        {
-                            return ActionPreference.Inquire;
-                        }
-                    }
-                    else
-                        return ActionPreference.SilentlyContinue;
-                }
+                    return Debug ? ActionPreference.Continue : ActionPreference.SilentlyContinue;
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/MshCommandRuntime.cs]: Please add open and closing braces.

**Review by PaulHigin — CHANGES_REQUESTED** [review]
I fully agree with this change.  To me -Debug is much more useful as a data stream rather than a prompt.

**Comment by iSazonov** [comment]
PowerShell-CI-Windows temporary failed. Restart it.

**Review by PaulHigin — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee the current PR precludes the ability to opt into the current behavior to use `Inquire`.  We agree that the default can be `Continue`, but we still need a way to allow users to use `Inquire`.  Suggestion is to add `-DebugAction` to allow going back to current behavior for a specific cmdlet, however, that can be deferred until someone needs it.

**Comment by KirkMunro** [comment]
If you happen to get such a request (someone asking for `Inquire` support), please tag me, I'd love to learn why someone wants that support.

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by iSazonov** [comment]
Is the PR ready to merge? If so @KirkMunro please update the PR description (breaking change is not mentioned).

**Comment by KirkMunro** [comment]
@iSazonov: It is ready to merge, and I added more descriptive details to the first paragraph in the description to identify what would be breaking.

**Comment by iSazonov** [comment]
@KirkMunro Thanks for the great contribution!

**Comment by JeremyTBradshaw** [comment]
> If you happen to get such a request (someone asking for `Inquire` support), please tag me, I'd love to learn why someone wants that support.

I am a little late jumping to PowerShell 7.1.  Recently started to while I am using MS Graph and EWS Managed API and have no need for Windows PowerShell when doing so.  I was caught off guard by this change, and am very used to -Debug resulting in Inquire.  I found debugging this way, way easier than anything else.  Debugging in VS Code seemed difficult and so I stuck with -Debug = Inquire.

I'm not necessarily opposed to the new way, but I would like to know where is some decent documentation on how to debug as easily.  I really like to Suspend, jump in and investigate.  It's how I hone in all my scripts and flush out errors.  I don't like to just see streams of info blowing by, so really need that Suspend capability.

Please do share if you can.  I'll also start researching on the topic.  This is a doozy of a change indeed.  If it means debugging the old way in a regular PowerShell console isn't possible, ouch.

**Comment by JeremyTBradshaw** [comment]
Actually, you know it's fine.  I will just get used to it.  I am a little lost on debugging in general so that's on me.  I see how -ErrorAction Break is essentially what I need for replacing the older -Debug = Inquire in Windows PowerShell.

Using ErrorAction Break for the first time, the top 1/2 of the instructions were covered over by my frozen Write-Progress.  And scrolling up, all the text is garbled.  Quickly put my tail between my legs and just went back to Windows PowerShell.  Saving this bit of learning for a future date.

**Comment by Samuli3682** [comment]
Jeah...i am totally out of what i'm trying to do...just want  working twrp.
to my mate Pro 10bla-l29....

to 3. jouluk. 2020 klo 5.31 Jeremy <notifications@github.com> kirjoitti:

**Comment by KirkMunro** [comment]
Hey @JeremyTBradshaw, thanks for responding back. I was really curious if there would be resistance among some folk, and the fact that you're responding rather than just silently complaining is very helpful.

I'm happy to help you out with this, either via back-and-forth texts, or a call on Microsoft Teams where we can screenshare and I could show you the ropes of the debugger. That would be helpful for me as well, to learn how you would like to use the debugger, and make sure that your needs are covered. Maybe the outcome of this will be a document describing what you're looking for: how to transition from `-Debug`'s Inquire capability to the using the full power of the debugger without losing your mind in the process.

**Comment by KirkMunro** [comment]
> Using ErrorAction Break for the first time, the top 1/2 of the instructions were covered over by my frozen Write-Progress. And scrolling up, all the text is garbled. Quickly put my tail between my legs and just went back to Windows PowerShell. Saving this bit of learning for a future date.

For this, might I suggest setting `$ProgressPreference = 'SilentlyContinue'`, and then running your script with `-ErrorAction Break`? That would prevent the progress window from appearing, which should help what you just tried a little. Anyway, if you want to do a call to discuss this and more, I'll be able to fast forward your learning quite a bit.

**Comment by KirkMunro** [comment]
> If it means debugging the old way in a regular PowerShell console isn't possible, ouch.

One last thing @JeremyTBradshaw: I do almost all of my debugging in a regular PowerShell console. Always have. Only rarely will I debug in an IDE like PowerShell ISE or VS Code. All to say, debugging in a regular PowerShell console works wonderfully.

You can also do old-school debugging in PowerShell 7 anytime you like. Simply set `$DebugPreference = 'Inquire'` and then run your script, and when you do, you'll get the prompts that you are used to; however, it would be beneficial to learn the debugger itself and to understand how to leverage what it can do for you -- you'll be more efficient when you need to debug scripts as a result.

**Comment by JeremyTBradshaw** [comment]
> > If it means debugging the old way in a regular PowerShell console isn't possible, ouch.
> 
> One last thing @JeremyTBradshaw: I do almost all of my debugging in a regular PowerShell console. Always have. Only rarely will I debug in an IDE like PowerShell ISE or VS Code. All to say, debugging in a regular PowerShell console works wonderfully.
> 
> You can also do old-school debugging in PowerShell 7 anytime you like. Simply set `$DebugPreference = 'Inquire'` and then run your script _without_ `-Debug`, and when you do, you'll get the prompts that you are used to; however, it would be beneficial to learn the debugger itself and to understand how to leverage what it can do for you -- you'll be more efficient when you need to debug scripts as a result.

Thanks for this.  I would take you up on the Teams meeting offer if it still stands.  Not sure when you had in mind but I've just exposed my email address on my GitHub profile and my tomorrow is currently open so if that works, that'd be great.  I should be quick to pick it up, so really it won't be long, and I can quickly show you the way I've been using -Debug.  The short version is that I commonly place Write-Debug's in my code wherever I'd like to jump in to troubleshoot or test, and then run with -Debug.  Whereas you're confirming you do your proper debugging in the console, that is good to hear, so again, look forward to the chance to get a quick rundown on this.

Thanks!

**Comment by KirkMunro** [comment]
@JeremyTBradshaw: Responded via the email address you shared, and we can work out a time through email.

**Pull request merged**